src/hotspot/share/opto/ifnode.cpp

```diff
@@ -837,8 +857,102 @@ bool IfNode::has_only_uncommon_traps(ProjNode* proj, ProjNode*& success, ProjNod
   return false;
 }
 
+// There might be an AddINode (marked with *) with a constant increment
+// in-between the CmpNode(s) and the common value we compare.
+// Check for the following cases. Then return common value compared
+// if present and also save the constant values to be adjusted or
+// subtracted due to the possible AddINode in-between.
+//
+//   Variant 1         Variant 2         Variant 3           Variant 4
+//
+//    res_val           res_val            res_val             res_val
+//     /   \             /   \             /   \               /     \
+// dom_cmp  \        dom_val* \           /  this_val*    dom_val*   this_val*
+//        this_cmp      |      \         /       \            |        \
+//                   dom_cmp    \     dom_cmp     \        dom_cmp      \
+//                            this_cmp          this_cmp              this_cmp
+Node* IfNode::get_base_comparing_value (Node* dom_if, PhaseIterGVN* igvn, jint& this_adj_val, jint& dom_adj_val) {
+  Node* this_cmp = in(1)->in(1);
+  Node* dom_cmp = dom_if->in(1)->in(1);
+  assert(this_cmp->is_Cmp() != false && dom_cmp->is_Cmp() != false, "compare expected");
+
+  Node* res_val = NULL;
+  this_adj_val = 0;
+  dom_adj_val = 0;
+  Node* dom_val = dom_cmp->in(1);
+  Node* this_val = this_cmp->in(1);
+  if (this_val == dom_val) {
+    res_val = this_val;
+  } else if (this_val->is_Add() && this_val->in(1) && this_val->in(1) == dom_val) {
+    const TypeInt* val_t = igvn->type(this_val->in(2))->isa_int();
+    if (val_t && val_t->is_con()) {
```

Comment:
`val_t` implicitly casts a pointer to a boolean, should be replaced by `val_t != NULL`. Same below.